### ISSUE TITLE
Create task to run the app

### ISSUE THREAD

**Issue by georgedouzas** [NOVA-IMS-Innovation-and-Analytics-Lab/MapIntel#81]
We should have a task to run the app, combining the commands in the README file. The app should start with a command like:

`pdm run`

 How does the app get the data? Probably the above command should be modified to point to the data:

`pdm run data.json`

**Comment by georgedouzas** [comment]
> We should have a task to run the app, combining the commands in the README file. The app should start with a command like:
> 
> `pdm run`
> 
> How does the app get the data? Probably the above command should be modified to point to the data:
> 
> `pdm run data.json`

@stuartgo

**Comment by stuartgo** [comment]
Sorry did not see this, currently, I have a notebook that sets up the database and feeds it the data. It relies on the other services as it calls the services to create the vectors to then store them in the database. As mentioned the database is on a free trial, which is time limited so I am not sure if it makes sense to use the script to fill the database when its started if the script does not also at the same time create a self hosted database

**Comment by georgedouzas** [comment]
For now, let's forget the current implementation. Can we design the system to accept data in a standardized format (or formats)?

**Comment by stuartgo** [comment]
I think it already is, for documents it uses a dict with these keys:
['document_id', 'timestamp', 'url', 'title', 'topic_label', 'umap_embeddings', 'image_url', 'snippet', 'topic_number']
and for topics it just needs the topic name with the key 'topic'

**Comment by stuartgo** [comment]
Ah yes, but I guess its the same except topic_label,umap_embeddings, and topic_number. or something similar to that

**Comment by georgedouzas** [comment]
Also I guess we would like to have:

1) The transformed data stored in a db.
2) The db should have a KNN retrieval functionality.

I think this already exists as db service in the current implementation, correct? Self-hosted is not critical here (you mean it calls elasticsearch hosted in a cloud platform?).

So we would also like to have an extra endpoint to the db service that populates the database using a POST request and the raw data. To populate the db should obviously call the other services.

**Comment by georgedouzas** [comment]
The idea is that given the raw data, we can call a service (that, in turn, utilizes the other services) to populate the database. In general, it does not matter where that database is hosted.

**Comment by stuartgo** [comment]
Should the database service then have endpoints to populate its database then?

**Comment by georgedouzas** [comment]
Can't we add an endpoint here?

https://github.com/NOVA-IMS-Innovation-and-Analytics-Lab/MapIntel/blob/master/src/mapintel/services/database/api/database.py

**Comment by georgedouzas** [comment]
This endpoint would be used as follows:

1. POST request with the raw data
2. It will call the other services to transform the data.
3. It will populate the database with the transformed data.

What do you think?

**Comment by stuartgo** [comment]
Yes that makes sense, I can try to make it now, I should have most of the code already

**Comment by georgedouzas** [comment]
> Ok does it need to be a service or can it just be a python script that fetches the data from a file?

Yes it can be also some client code but still the database needs some internal mechanism to be updated (an endpoint basically) or else it is like an external service to mapintel.

**Comment by georgedouzas** [comment]
> Yes that makes sense, I can try to make it now, I should have most of the code already

Perfect. Then you can start from an empty database and test that it gets populated with some data. Finally we can add some tasks to populate the database, start the app, check status etc

**Comment by georgedouzas** [comment]
All these are important because we will need at some point to change data and it should be a fast process.

**Comment by stuartgo** [comment]
True. The service would also need to fit all the models to that data though, right? So it would be:

1. POST request with the raw data
2. It will train models on given data
3. It will call the other services to transform the data.
4. It will populate the database with the transformed data.

Or can we assume that the user fits the models to the data and updates them in mapintel?

**Comment by georgedouzas** [comment]
The process you described should be done through a client script that:

1. Trains the models using the raw data.
2. Populates that elasticsearch database with the transformed data (by POSTing the raw data).

This functionality should be supported by the underlying services. I think only 2 is missing right now?

Ideally these services should be decoupled. My suggestion violates because db service calls other servicess too. So maybe we should POST not the raw data but the transformed?

**Comment by georgedouzas** [comment]
Then everything is decoupled and the client script will have the responsibility to call the other services in the correct order  such that the system works as expected.

**Comment by georgedouzas** [comment]
We need to provide all this scripts from the CLI as named tasks. Let's look at a some scenarios and tasks:

1. Elasticsearch is empty and the user has some raw data (basically a list of documents i.e. a list of strings?). The user calls `pdm run` to start the app. An error should inform the user that mapintel has not yet been initialized. On the other hand, If it is already initialized, then the command should start the app.
2. To initialize the db we will need a command like `pdm init data`. This command should call a script that 1) Trains the models 2) Populates the database
4. We might need also some tasks to update the database and retrain the models but let's skip it for now.

**Comment by georgedouzas** [comment]
> Yes I think that might make more sense, so essentially the db endpoint would just have to store the data it is given?

Yes, but this is the function when we POST transformed data. We might need also an UPDATE functionality (or GET to get some info - I think you alredy support through another endpoint?)

**Comment by georgedouzas** [comment]
I am confused. All the above can be done offline. How is inference functionality been used on the runtime of the app?

**Comment by stuartgo** [comment]
I think the only inference done during runtime is the query field, which essentially just plots a single point that has to be vectorized and then reduced to two dimensions.

**Comment by georgedouzas** [comment]
> I think the only inference done during runtime is the query field, which essentially just plots a single point that has to be vectorized and then reduced to two dimensions.

Ok thanks! But this is also important.

Another question is if the elasticsearch stored data (the raw data actually) and the training data have to be the same.

**Comment by stuartgo** [comment]
``

> > Yes I think that might make more sense, so essentially the db endpoint would just have to store the data it is given?
> 
> Yes, but this is the function when we POST transformed data. We might need also an UPDATE functionality (or GET to get some info - I think you alredy support through another endpoint?)

What kind of info do you mean here?

**Comment by georgedouzas** [comment]
> ``
> 
> > > Yes I think that might make more sense, so essentially the db endpoint would just have to store the data it is given?
> > 
> > 
> > Yes, but this is the function when we POST transformed data. We might need also an UPDATE functionality (or GET to get some info - I think you alredy support through another endpoint?)
> 
> What kind of info do you mean here?

Retrieve the stored data and the corresponding documents.

**Comment by stuartgo** [comment]
Ok then yes there are endpoints to do that

**Comment by stuartgo** [comment]
> asticsearch stored data (the raw

Good point, maybe it does not make sense to include a  `pdm data init` that trains and populates with the same data then?

**Comment by georgedouzas** [comment]
> > asticsearch stored data (the raw
> 
> Good point, maybe it does not make sense to include a `pdm data init` that trains and populates with the same data then?

Well, initially you need both of them. We might want to change it to `pdm init --train data1 --db data2` in case that the user wants to provide different data. I dont know if this is useful or not. What do you think?

**Comment by stuartgo** [comment]
yes I think so, it allows for more flexibility

**Comment by stuartgo** [comment]
I need to leave for practice soon, but ill try to repurpose my code into a script that does the init sequence tomorrow.

**Comment by stuartgo** [comment]
> > > asticsearch stored data (the raw
> > 
> > 
> > Good point, maybe it does not make sense to include a `pdm data init` that trains and populates with the same data then?
> 
> Well, initially you need both of them. We might want to change it to `pdm init --train train_data --db db_data` in case that the user wants to provide different data. I dont know if this is useful or not. What do you think?

Just to make sure I got it right, this would assume that the models have already been loaded into mapintel?

**Comment by stuartgo** [comment]
But then again I think earlier we assumed that the models would be loaded pretrained so then the point of training disappears a bit

**Comment by georgedouzas** [comment]
The general idea is that we will to implement some client code (in the form of nox tasks). This client code will interact with the already implemented services. So the services should support what I will describe shortly. If not, then we will need to implement it. Please comment on the following:

1. A task that starts the app: `pdm start`. If the app is initialized then this will work as expected, if not then it should return an appropriate error msg.
2. A task that initializes the app: `pdm init ....` . This command also accepts arguments related to the data, maybe as described above. The task should apply all the required steps to initialize the app i.e. populate the database, populate the model registry with default models (if they are empty), train them etc. All these should be supported functionality by the services. After this command is called then `pdm start` should work with no other requirement.
3. Tasks to update the app: This might include updating the db or retrain a selected model or select another model etc
4. Tasks to monitor the app: Information about the current status of db, selected models etc

**Comment by georgedouzas** [comment]
So the tasks will hide from the user the complexity and the various steps required to set up the system. Internally they will interact with the services with the right order etc

**Comment by georgedouzas** [comment]
I propose to start with `pdm start` assuming that the app is initialized. Once this is done then we will create an empty state of the app and make sure we get the correct error msg. Then we can move forward with `pdm init ...`

**Comment by georgedouzas** [comment]
Btw these tasks are there for convinience. The user might choose to interact with the services in any eay (not reccomended though for complex tasks). For example the mlflow service provides some endpoints. Also mlflow provides a native server and a GUI to interact.

**Comment by stuartgo** [comment]
Ok with the default models I think it makes sense. I have never made a task before. But for `pdm start` it would essentially just call the two commands I added in the README. Ill make a script for `pdm init`

**Comment by georgedouzas** [comment]
I would suggest to start with `pdm start`. There are two steps:

1. Add a task to the `noxfile.py`
2. Add a script command to `pyproject.toml`

The task should deal also with the case that the app needs to be initialized with an appropriate error msg.

**Comment by stuartgo** [comment]
I dont really know the syntax, but does this look ok?

`start = "nox --error-on-external-run -R -s start"`

and then I add a @nox.session decorator to the function I want ran?

**Comment by georgedouzas** [comment]
Yes!

Basically this will go to `pyproject.toml` and define the `pdm` script that can be invoked with the command `pdm start`.

This command in turn will invoke nox session called start i.e. a decorated function that you will create.

**Comment by georgedouzas** [comment]
Follow https://github.com/NOVA-IMS-Innovation-and-Analytics-Lab/MapIntel/blob/master/noxfile.py#L29

**Comment by stuartgo** [comment]
Ok nice that worked, do I need to configure nox with multiple interpreters? I run into this error.
`RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.10'`

**Comment by georgedouzas** [comment]
For this particular task you just need a single python version (already selected in the decorator). Nox will try to install mapintel in a virtual environment but it needs an available python interprenter to do it. It checks a few common path. Do you have python 3.10 installed?

**Comment by stuartgo** [comment]
yes, but through asdf I think. I get this error when trying to open python shell, but couldn't find what to do.
`No version is set for command python
Consider adding one of the following versions in your config file at
python 3.9.9
python 3.10.1`

**Comment by georgedouzas** [comment]
try `asdf shell python 3.10.1`

**Comment by stuartgo** [comment]
Doesnt seem to do anything, is there anyway I can make nox detect the asdf python versions?

**Comment by georgedouzas** [comment]
Let's do something else, since we can use nox only for development tasks. Add a file `cli.py` into the `src/mapintel. This will implement the CLI. We can use [click](https://click.palletsprojects.com/en/8.1.x/) to define commands like `mapintel start` or `mapintel init` etc

**Comment by stuartgo** [comment]
This is only to feed the topic names into the database though, so maybe we can just leave that up to the users?

**Comment by georgedouzas** [comment]
I think the topic names should be added by the user when the app is initialized.

**Comment by stuartgo** [comment]
But arent the topic names somewhat dependent on the behaviour of the topic modelling part, so therefore also dependent of the training. So the user might not know the topic names before the training is done?

**Comment by georgedouzas** [comment]
I am not sure. Can we at least constraint the number of topics for the various topic modeling methods? If yes, then it makes sense to define the topics before defining or training the models.

**Comment by stuartgo** [comment]
Yes i think you can define the number in most models. As it is now the topic names are extracted after the model is trained though, I think that is the most common way to do it. As it is now the topic names are essentially the words that most commonly occur in a "cluster" of articles. So I don't see how you can define the topic names before training, except maybe if we just use numbers?

**Comment by georgedouzas** [comment]
Ok I see. So assume that you can get them from the model and we will see how to generalize.

**Comment by stuartgo** [comment]
I ran into some problems with the wrapper of the topic modelling so its not working 100%, but I have to leave soon, should I push it so you can have a look or leave it for tomorrow?

**Comment by georgedouzas** [comment]
Yes!

**Comment by stuartgo** [comment]
Just pushed now, the script should work, I was able to fix the issue. Its not really connected to anything though

**Comment by georgedouzas** [comment]
Please open a new PR.
